Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,156 @@
+    /// How to connect to the host.
+    public var command: Command
+    
+    /// The target host address.
+    public var addressType: AddressType
+    
+    /// The target host port.
+    public var desiredPort: UInt16
+    
+    /// Creates a new `ClientRequest`.
+    /// - parameter command: How to connect to the host.
+    /// - parameter addressType: The target host address.
+    /// - parameter desiredPort: The target host port.
+    public init(command: Command, addressType: AddressType, desiredPort: UInt16) {
+        self.command = command
+        self.addressType = addressType
+        self.desiredPort = desiredPort
+    }
+    
+}
+
+extension ByteBuffer {
+    
+    @discardableResult mutating func writeClientRequest(_ request: ClientRequest) -> Int {
+        var written = 0
+        written += self.writeInteger(request.version)
+        written += self.writeInteger(request.command.value)
+        written += self.writeInteger(0, as: UInt8.self)
+        written +=  self.writeAddressType(request.addressType)
+        written += self.writeInteger(request.desiredPort)
+        return written
+    }
+    
+}
+
+// MARK: - Command
+
+/// What type of connection the SOCKS server should establish with
+/// the target host.
+struct Command: Hashable {
+    
+    /// Typically the primary connection type, suitable for HTTP.
+    static let connect = Command(value: 0x01)
+    
+    /// Used in protocols that require the client to accept connections
+    /// from the server, e.g. FTP.
+    static let bind = Command(value: 0x02)
+    
+    /// Used to establish an association within the UDP relay process to
+    /// handle UDP datagrams.
+    static let udpAssociate = Command(value: 0x03)
+    
+    var value: UInt8
+}
+
+// MARK: - AddressType
+
+/// The address used to connect to the target host.
+public enum AddressType: Hashable {
+    
+    /// An IPv4 address (4 bytes), e.g. *192.168.1.2*
+    case ipv4([UInt8])
+    
+    /// A fullly-qualified domain name, e.g. *apple.com*
+    case domain([UInt8])
+    
+    /// An IPv6 address (16 bytes), e.g. *aaaa:bbbb:cccc:dddd*
+    case ipv6([UInt8])
+    
+    /// How many bytes are needed to represent the address
+    var size: Int {
+        switch self {
+        case .domain(let domain):
+            return domain.count + 1
+        case .ipv4:
+            return 4
+        case .ipv6:
+            return 16
+        }
+    }
+    
+    init?(buffer: inout ByteBuffer) throws {
+        guard let type = buffer.readInteger(as: UInt8.self) else {
+            return nil
+        }
+        
+        switch type {
+        case 0x01:
+            guard let bytes = buffer.readBytes(length: 4) else {
+                return nil
+            }
+            self = .ipv4(bytes)
+        case 0x03:
+            guard
+                let length = buffer.readInteger(as: UInt8.self),
+                let bytes = buffer.readBytes(length: Int(length))
+            else {
+                return nil
```

Comment:
- reset of buffer missing here
- usually we implemented these methods as extensions on `ByteBuffer`. ie `mutating func readAddressType()`

see here: https://github.com/apple/swift-nio-extras/blob/main/Tests/NIOExtrasTests/WritePCAPHandlerTest.swift#L716-L845

Reply:
Race condition, just done